rfcs/0191-worker-manager-launch-configs.md

```diff
@@ -0,0 +1,122 @@
+# RFC 0191 - Worker Manager launch configurations
+* Comments: [#0191](https://github.com/taskcluster/taskcluster-rfcs/pull/191)
+* Proposed by: @lotas
+
+## Summary
+
+This proposal outlines enhancements to the worker-manager service that would address the following:
+
+* make worker pools launch configurations first-class citizens in the worker-manager service
+* balanced distribution of workers across multiple launch configurations (regions)
+* dynamic adjustment of launch configuration likelihood to be provisioned based on **health metrics**
+
+**Health metrics** - is the combination of the number of fail-to-success provisioning attempts ratio, the number of workers currently running.
+
+## Motivation
+
+The current worker-manager lacks the ability to deal with broken configurations, monitor the health of launch configurations and the availability of cloud resources.
+
+This often leads to the creation of workers that are stuck in provisioning or fail to become operational, wasting resources and reducing system efficiency.
+
+By introducing health tracking and dynamic adjustment mechanisms, we aim to optimize the provisioning process and enhance system reliability.
+
+## Details
+
+New entity will be introduced to worker-manager - **Launch Configuration**.
+Each worker pool will have zero or more **launch configurations** associated with it.
+
+Each **launch configuration** will be assigned a unique ID (hash of its properties), stored in the database and will be immutable (except for the status flag).
+
+Generated ID will include the following properties:
+
+* top-level config properties (`minCapacity`, `maxCapacity`, `scalingRatio`, `lifecyle`)
+* launch configuration itself
+
+During updates of the worker pool configurations, all existing launch configurations that are not present in the new configuration will be marked as `archived`. Launch configurations that have same unique ID (hash) will be kept active. All other will be created as new launch configurations and marked as active.
+
+### Launch Configuration weight
+
+Provisioned workers would be associated with a specific launch configuration (`worker.launchConfigurationId`).
+This will allow us to to know how many workers with this launch configuration were successfully registered and claimed work.
+
+Each launch configuration will have a dynamic `weight` property that will be adjusted automatically based on the following events and metrics:
+
+* total number of successful worker provisioning attempts / registrations
+* total number of failed worker provisioning attempts / registrations
+* has any worker claimed a task
+* fail-to-success ratio over a specific time period (i.e. last hour)
+* number of non-stopped workers currently running
+
+Weight will be used to determine the likelihood of selecting a specific launch configuration for provisioning.
```

Comment:
When we talked about this sort of health-monitoring in the context of aws-provisioner, jhford noted that if we _never_ try a configuration, we'll have no way to know if its status has changed (region back online, etc). I suspect weights should be constrained to be strictly >0.

Reply:
yes yes, I kept that in mind. We don't want to fully "punish" the region, since most of those errors could/should be temporary. 

I also think that we should narrow the scope of this RFC to the "weights" only, without touching the provisioning rate. I think I've tried to cover both issues - even region distribution and dealing with broken configs. 

Should there be a separate RFC for dealing with new launch configs after this one lands? 
Maybe we could use the existing `scalingRatio` once this gets implemented, that would take new metrics into account and lower that value for totally new configs.